src/appleseed/renderer/modeling/bsdf/microfacetbrdfhelper.h

```diff
@@ -0,0 +1,173 @@
+
+//
+// This source file is part of appleseed.
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2010-2013 Francois Beaune, Jupiter Jazz Limited
+// Copyright (c) 2014-2015 Francois Beaune, The appleseedhq Organization
+// Copyright (c) 2014-2015 Esteban Tovagliari, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// appleseed.renderer headers.
+#include "renderer/global/globaltypes.h"
+#include "renderer/modeling/bsdf/bsdfsample.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/basis.h"
+
+// Standard headers.
+#include <algorithm>
+#include <cmath>
+
+namespace renderer
+{
+
+template <typename T>
+class MicrofacetBRDFHelper
+{
+  public:
+
+    typedef foundation::Vector<T, 3>    VectorType;
+    typedef foundation::Basis3<T>       BasisType;
+
+    template <typename MDF, typename FresnelFun>
+    static void sample(
+        const MDF&  mdf,
+        const T     alpha_x,
+        const T     alpha_y,
+        const T     g_alpha_x,
+        const T     g_alpha_y,
+        FresnelFun  f,
```

Comment:
I'm ok with f. It's a function object and the type makes that clear.
I did the other fixes.